embassy-net/src/tcp.rs

```diff
@@ -359,6 +417,55 @@ impl<'d> TcpIo<'d> {
         .await
     }
 
+    async fn send<F, R>(&mut self, f: F) -> Result<R, Error>
+    where
+        F: FnOnce(&mut [u8]) -> (usize, R),
+    {
+        let mut f = Some(f);
+        poll_fn(move |cx| {
+            self.with_mut(|s, _| {
+                if s.send_queue() == s.send_capacity() {
+                    // socket buffer is full wait until it has atleast one byte free
+                    s.register_send_waker(cx.waker());
+                    Poll::Pending
+                } else {
+                    Poll::Ready(match s.send(f.take().unwrap()) {
+                        // Connection reset. TODO: this can also be timeouts etc, investigate.
+                        Err(tcp::SendError::InvalidState) => Err(Error::ConnectionReset),
+                        Ok(r) => Ok(r),
+                    })
+                }
+            })
+        })
+        .await
+    }
+
+    async fn recv<F, R>(&mut self, f: F) -> Result<R, Error>
+    where
+        F: FnOnce(&mut [u8]) -> (usize, R),
+    {
+        let mut f = Some(f);
+        poll_fn(move |cx| {
+            self.with_mut(|s, _| {
+                if s.recv_queue() == 0 {
```

Comment:
you should do more checks here, so that "This function errors if the receive half of the connection is not open." is true. As it is now it'll hang instead of erroring.